_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+
+Następnie definiujemy sam efekt:
+* zaglądamy do strumienia akcji 
+```typescript 
+this.actions$.pipe()
+```
+* szukamy tylko akcji określonego typu 
+```typescript
+ofType(getActivityType)
+```
+* reagujemy na każdą z nich poprzez wywołanie żądania HTTP i zaglądamy do strumienia zawierającego odpowiedź z API 
+```typescript 
+mergeMap(() => this.http.get<ActivityItemModel>('https://www.boredapi.com/api/activity').pipe()
+```
+* mapujemy odpowiedź na wywołanie akcji `map(response => (addActivity(response)))`
+```typescript 
+map(response => (addActivity(response)))
+```
+* w przypadku błędu zwracamy pustą wartość Obervable
+```typescript 
+catchError(() => EMPTY)
+```
+
+W wyniku działania tego efektu każde wywołanie akcji `Get Activity` przy poprawnej odpowiedzi z API wywoła akcję `Add Activity` z otrzymaną czynnością.
+
+### Reducer
+W reducerach zazwyczaj znajduje się najwięcej logiki ze wszystkich komponentów NgRx. To tutaj trzeba zdecydować jak dodawać do store kolejne dane. Warto pamiętać, że reducery to **czyste funkcje**, które przyjmują jako parametr akcję i obecny stan aplikacji, a następnie zwracają nowy stan, dzięki czemu zachowujemy jego niemutowalność. W naszym przypadku wygląda to tak:
+
+```typescript
+const initialState: ActivityItemsState = {activities: []};
+
+export const activityReducer = createReducer(
+    initialState,
+    on(addActivity, (state, payload) => ({
+        ...state,
+        activities: state.activities.concat({activity: payload.activity, participants: payload.participants} as ActivityItemModel)
+    }))
+);
+```
+
+W pierwszej linii definiujemy jakiego stanu początkowego store (lub - jak ma to miejsce powyżej - jego wycinka) się spodziewamy. W naszym przypadku będzie to pusta tablica. Następnie wywołujemy funkcję `createReducer`, która w pierwszym parametrze przyjmuje wspomniany stan początkowy, a w drugim (i każdym kolejnym) - definicję odpowiedzi na przybyłą akcję. W tym przypadku reducer reaguje na akcję `Add Activity` poprzez dodanie do dotychczasowej tablicy `activities` nowego elementu i zwrócenie całości jako wynik wywołania funkcji anonimowej. Reducer jest częścią całej architektury, którą można w bardzo łatwy sposób przetestować i zapewnić sobie w ten sposób pewność jego poprawnego działania.
+
+### Selektor
+Ostatnią częścią układanki w przepływie są selektory. Podobnie jak reducery są one czystymi funkcjami, których zadaniem jest obserwowanie wycinków store i dostarczanie informacji o ich zmianach do komponentów. W naszej aplikacji selektor zdefiniowany jest następująco:
```

Comment:
nie tylko, jest też coś ważniejszego, chociaż nie tak oczywistego na początku. selektory służą też to wyliczania stanu pochodnego po store i są miejscem, w którym mimo deduplikacji danych w store możemy je wygodnie i przyjemnie oglądać w komponentach. Jeśli zakładamy, że redux pozwala pisać komponenty tak jakby były w izolacji i nie wiedziały nic o ngrx to selektory pozwalają wyciągać przyjazne dla nich projekcje/widoki bieżącego stanu świata (a akcje zabierają od nich aspekt interakcji, co jako całość pozwala pisać "głupie" komponenty)